src/block_helpers.rs

```diff
@@ -1,59 +1,82 @@
 use std::error::Error;
 use crate::optimizer::OptimizerTrait;
 use std::io;
+use std::io::{Read};
+
 use std::slice;
-use std::mem::{self, MaybeUninit};
+use std::mem::{self};
 use std::cmp::min;
-use std::ops::Deref;
 use crate::optimizer::OptimizerSGD;
-use std::marker::PhantomData;
 use crate::feature_buffer;
 use crate::regressor::BlockTrait;
+use crate::port_buffer;
+use crate::graph;
 
 #[derive(Clone, Debug)]
 #[repr(C)]
 pub struct Weight {
     pub weight: f32, 
 }
 
+#[derive(Clone, Debug)]
+#[repr(C)]
+pub struct OptimizerData<L:OptimizerTrait> {
+    pub optimizer_data: L::PerWeightStore,
+}
+
 #[derive(Clone, Debug)]
 #[repr(C)]
 pub struct WeightAndOptimizerData<L:OptimizerTrait> {
     pub weight: f32, 
     pub optimizer_data: L::PerWeightStore,
 }
 
+
```

Comment:
It appears [](https://docs.rs/assert_float_eq/latest/assert_float_eq/) crate is used for approximate assertions, might be useful down the road